Subdiv Corners/Creases in Houdini (and IECoreScene::MeshAlgo)

andrewkaufman opened the pull request:
This should finish off our push to support subdiv corners/creases in Cortex. I've:

- Added corners/crease import/export for Houdini polygons
- Fixed a bug in the mesh interpolation export from Houdini
- Updated `MeshAlgo::deleteFaces`, `MeshAlgo::triangulate`, `MeshAlgo::reorderVertices` to maintain the corners/creases
- Ported `MergeMergeOp` to `MeshAlgo::merge` and updated it to maintain the corners/creases

johnhaddon commented on src/IECoreScene/MeshAlgoMerge.cpp:
```diff
@@ -0,0 +1,297 @@
+	/// \todo: can this be parallelized?
+	MeshPrimitivePtr result = meshes[0]->copy();
+	auto it = meshes.begin() + 1;
+	for( ; it != meshes.end(); ++it )
```
I suspect this scales poorly with increasing numbers of meshes. Rather than allocating enough memory for everything and filling it once, we're re-allocating and re-copying from the start for every single mesh added - I just discovered that this is called [Schlemiel the Painter's Algorithm](https://en.wikichip.org/wiki/schlemiel_the_painter%27s_algorithm), and it even has its own joke! Since this isn't worse than the previous option of doing the same manually via MeshMergeOp, and I suspect you want to move on to other tasks, I suggest we change the comments about parallelisation to a comment about addressing this problem instead.

andrewkaufman replied:
Todo updated in 5e25f85

andrewkaufman commented on src/IECoreScene/MeshAlgoMerge.cpp:
```diff
@@ -0,0 +1,297 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
```
Fixed in 5e25f85

andrewkaufman commented on src/IECoreHoudini/ToHoudiniPolygonsConverter.cpp:
```diff
@@ -41,6 +41,16 @@ using namespace IECore;
+namespace
+{
+
+static InternedString g_interpolationAttrib( "ieMeshInterpolation" );
```
Fixed in 037238d

andrewkaufman commented on src/IECoreHoudini/ToHoudiniPolygonsConverter.cpp:
```diff
@@ -110,8 +120,8 @@ bool ToHoudiniPolygonsConverter::doConversion( const Object *object, GU_Detail *
 	{
-		std::string interpolation = ( mesh->interpolation() == "catmullClark" ) ? "subdiv" : "poly";
-		ToHoudiniStringVectorAttribConverter::convertString( "ieMeshInterpolation", interpolation, geo, newPrims );
+		std::string interpolation = ( g_catmullClark == mesh->interpolation() ) ? g_subdiv : g_poly;
```
Fixed in 037238d

andrewkaufman commented on src/IECoreHoudini/FromHoudiniPolygonsConverter.cpp:
```diff
@@ -188,5 +190,111 @@ ObjectPtr FromHoudiniPolygonsConverter::doDetailConversion( const GU_Detail *geo
+
+	IntVectorDataPtr cornerIdsData = new IntVectorData();
+	auto &cornerIds = cornerIdsData->writable();
+	// likely larger than necessary, but we don't know the correct size yet
```
Fixed in 2c04571

andrewkaufman commented on src/IECoreScene/MeshAlgoDeleteFaces.cpp:
```diff
@@ -49,6 +49,92 @@ using namespace IECoreScene;
+
+	IntVectorDataPtr outIdData = new IntVectorData;
+	auto &outIds = outIdData->writable();
+	// potentially too large but large enough
```
Fixed in 780c5a1

andrewkaufman commented on src/IECoreScene/MeshAlgoMerge.cpp:
```diff
@@ -249,8 +249,57 @@ void merge( MeshPrimitive *a, const MeshPrimitive *b )
+		auto &ids = idData->writable();
+		ids.resize( aCornerIds.size() + bCornerIds.size() );
+		it = std::copy( aCornerIds.begin(), aCornerIds.end(), ids.begin() );
+		std::transform( bCornerIds.begin(), bCornerIds.end(), it, std::bind2nd( std::plus<int>(), vertexIdOffset ) );
```
Fixed in 3518716

andrewkaufman commented:
I've addressed all comments except one, rebased on master, and added one last commit to fixup some Houdini tests that were failing on master (since the addition of normals and uvs to the box).

The one comment remaining is indeed valid, but I'm not sure how to best solve it, or if its worth the effort right now, as it doesnt' come up in practice. I'll try to render that mesh to see what OpenSubdiv & Arnold do with those invalid creases...

andrewkaufman commented on src/IECoreScene/MeshAlgoDeleteFaces.cpp:
```diff
@@ -49,6 +49,92 @@ using namespace IECoreScene;
+			int id = remapping[ ids[creaseIdOffset + j] ];
+			if( id != -1 )
+			{
+				outIds.push_back( id );
```
I couldn't get the exact example working (planes don't seem to subdivide at all), but I tried something similar with a cube, creasing 0,1,2 and then deleting faces 0,1 which leaves a single orphaned crease edge (1-2 does not exist). In both OpenSubdiv and Arnold the result is the same:

![creaseCube](https://user-images.githubusercontent.com/3979682/54724503-aa554200-4b28-11e9-9ad5-e561b15e9773.png)

I'm not convinced that is correct, but I'm also not convinced its incorrect, and they're doing the same thing, so.... I'm inclined to merge and worry about it later when someone actually attempts this in practice.

andrewkaufman replied:
Todo added in fe2044ad2a12f68221d21dff0d29a11653cbde8b
